src/analysis/normed_space/M_structure.lean

```diff
@@ -0,0 +1,439 @@
+  ... ≥ ∥Q x∥ + 2 • ∥ (P * Q) x - (Q * P * Q) x∥ :
+    by exact add_le_add_right (norm_le_insert' (Q x) (Q (P (Q x)))) (2•∥(1-Q) (P (Q x))∥),
+  rw ge at e1,
+  nth_rewrite_rhs 0 ← add_zero (∥Q x∥) at e1,
+  rw [add_le_add_iff_left, two_smul,  ← two_mul]  at e1,
+  rw le_antisymm_iff,
+  split,
+  { rw ← mul_zero (2:ℝ) at e1,
+    rw mul_le_mul_left at e1, exact e1, norm_num, },
+  { apply norm_nonneg, }
+end
+
+lemma QP_eq_QPQ (P Q : X →L[𝕜] X) (h₁: is_Lprojection P) (h₂: is_Lprojection Q) : Q * P = Q * P * Q
+  :=
+begin
+  have e1: P * (1 - Q) = P * (1 - Q) - (Q * P - Q * P * Q) :=
+  calc P * (1 - Q) = (1 - Q) * P * (1 - Q) : by rw PQ_eq_QPQ P (1 - Q) h₁ h₂.Lcomplement
+  ... = 1 * (P * (1 - Q)) - Q * (P * (1 - Q)) : by {rw mul_assoc, rw sub_mul,}
+  ... = P * (1 - Q) - Q * (P * (1 - Q)) : by rw one_mul
+  ... = P * (1 - Q) - Q * (P - P * Q) : by rw [mul_sub, mul_one]
+  ... = P * (1 - Q) - (Q * P - Q * P * Q) : by rw [mul_sub Q, mul_assoc],
+  rw [eq_sub_iff_add_eq, add_right_eq_self, sub_eq_zero] at e1,
+  exact e1,
+end
+
+lemma Lproj_commute {P Q: X →L[𝕜] X} (h₁: is_Lprojection P) (h₂ : is_Lprojection Q) : commute P Q :=
+begin
+  unfold commute,
+  unfold semiconj_by,
+  rw PQ_eq_QPQ P Q h₁ h₂,
+  nth_rewrite_rhs 0 QP_eq_QPQ P Q h₁ h₂,
+end
+
+lemma mul {P Q : X →L[𝕜] X} (h₁ : is_Lprojection P) (h₂ : is_Lprojection Q) :
+  is_Lprojection (P*Q) :=
+begin
+  unfold is_Lprojection,
+  split,
+  { apply is_projection.mul_of_commute (Lproj_commute h₁ h₂) h₁.left h₂.left, },
+  { intro x,
+    rw le_antisymm_iff,
+    split,
+    -- rw map_sub, apply norm_add_le,
+    { calc ∥ x ∥ = ∥(P * Q) x + (x - (P * Q) x)∥ : by abel
+      ... ≤ ∥(P * Q) x∥ + ∥ x - (P * Q) x ∥ : by apply norm_add_le
+      ... = ∥(P * Q) x∥ + ∥(1 - P * Q) x∥ : rfl },
+    { calc ∥x∥ = ∥Q x∥ + ∥(1-Q) x∥ : by rw h₂.right x
+      ... = ∥P(Q x)∥ + ∥(1-P)(Q x)∥ + ∥(1-Q) x∥ : by rw h₁.right (Q x)
+      ... = ∥P(Q x)∥ + ∥Q x - P (Q x)∥ + ∥x - Q x∥ : rfl
+      ... = ∥P(Q x)∥ + (∥Q x - P (Q x)∥ + ∥x - Q x∥) : by rw add_assoc
+      ... ≥ ∥P(Q x)∥ + ∥(Q x - P (Q x)) + (x - Q x)∥ :
+        by apply (add_le_add_iff_left (∥P(Q x)∥)).mpr (norm_add_le (Q x - P (Q x)) (x - Q x))
+      ... = ∥P(Q x)∥ + ∥x - P (Q x)∥ : by rw sub_add_sub_cancel'
+      ... = ∥(P * Q) x∥ + ∥(1 - P * Q) x∥ : rfl }, }
+end
+
+lemma join {P Q: X →L[𝕜] X} (h₁ : is_Lprojection P) (h₂ : is_Lprojection Q) :
+  is_Lprojection (P + Q - P * Q) :=
+begin
+  have e1:  1 - (1 - P) * (1 - Q) = P + Q - P * Q :=
+  calc 1 - (1 - P) * (1 - Q) = 1 -(1 - Q - P * (1 - Q)) : by rw [sub_mul, one_mul]
+  ... = Q + P * (1 - Q) : by rw [sub_sub, sub_sub_self]
+  ... = P + Q - P * Q : by rw [mul_sub, mul_one, add_sub, add_comm],
+  rw ← e1,
+  rw ← is_Lprojection.Lcomplement_iff,
+  apply is_Lprojection.mul,
+  apply is_Lprojection.Lcomplement h₁,
+  apply is_Lprojection.Lcomplement h₂,
+end
+
+instance : has_compl(subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :=
```

Comment:
```suggestion
instance : has_compl (subtype (is_Lprojection  : (X →L[𝕜] X) → Prop)) :=
```

Reply:
and below.

Reply:
Does this mean that one should have e.g.
```
@[simp] lemma coe_compl (P : {P : X →L[𝕜] X // is_Lprojection P}) :
  ↑(Pᶜ) = (1:X →L[𝕜] X) - ↑P := rfl
```
Is there any way of making the inner `P` anonymous?